### PR TITLE
Add release notes for 0.246

### PR DESCRIPTION
**Pull request by caithagoras**
# Missing Release Notes
## Bhavani Hari
- [x] https://github.com/prestodb/presto/pull/15499 Add release notes for 0.245 (Merged by: Wenlei Xie)

## Emy Sun
- [ ] https://github.com/prestodb/presto/pull/15508 Port reverse(varbinary) from Prestosql to Prestodb (Merged by: Rongrong Zhong)

## George Wang
- [x] https://github.com/prestodb/presto/pull/14915 Avoid planning unnecessary LIMIT/TopN/Sort/DistinctLimit (Merged by: Rebecca Schlussel)
- [x] https://github.com/prestodb/presto/pull/14915 Avoid planning unnecessary LIMIT/TopN/Sort/DistinctLimit (Merged by: Rebecca Schlussel)

## shenh062326
- [x] d2cb24cd4cec630940463d8d8dc25806c774f00d Fix temporary table deletion for exchange materialization

# Extracted Release Notes
- #15322 (Author: Rebecca Schlussel): Only allow constants for prepared statement parameters
  - Fix an issue where prepared statements would allow some non-constant parameters.
- #15322 (Author: Rebecca Schlussel): Only allow constants for prepared statement parameters
  - Fix an issue where prepared statements would allow some non-constant parameters.
- #15470 (Author: George Wang): Ensure nullable partitioning column is handled properly
  - Fix dynamic pruning failures for joining on null keys in hive partition.
- #15485 (Author: Sreeni Viswanadha): Disallow ORDER BY literal in window functions
  - Disallow ORDER BY literals used with Window functions as it's not useful, expensive and most often used wrongly.
- #15501 (Author: Rongrong Zhong): Inject null provider when thrift execution is not needed
  - Fix injection issue when using function namespace manager.
- #15505 (Author: Saksham Sachdev): Add listener model for TaskThresholdMemoryRevokingScheduler
  - Add listener-based revocation model for spilling strategy PER_TASK_MEMORY_THRESHOLD.
- #15529 (Author: Rebecca Schlussel): Split large pages in TempSingleStreamSpiller
  - Fix a possible int overflow error when spilling to temp storage.
- #15550 (Author: Shixuan Fan): Support driver level plan in fragment result caching
  - Support fragment result caching for queries with local exchange (e.g. intermediate aggregation).
- #15550 (Author: Shixuan Fan): Support driver level plan in fragment result caching
  - Support fragment result caching for queries with local exchange (e.g. intermediate aggregation).
- #15555 (Author: Rongrong Zhong): Fix QualifiedObjectName instantiation in FunctionAndTypeManager
  - Fix a bug introduced in :pr:`15313` that would cause queries to fail when using upper case in SQL function catalog schema names.
- #15560 (Author: James Gill): Upgrade JTS to 1.18.0
  - Upgrade JTS to 1.18.0.
- #15585 (Author: Ariel Weisberg): Require min 30s query.min-expire-age
  - Add 30 second minimum to `query.min-expire-age` configuration parameter.
- #15592 (Author: Rebecca Schlussel): Disable spill for probe only grouped execution
  - Disable spill to disk for join queries where the probe side uses grouped execution and the build does not.  Previously these queries would fail with generic_internal_errors.

# All Commits
- dc00b9ff1df66f932326194b1303e677a5278f04 Disable spill for probe only grouped execution (Rebecca Schlussel)
- 8148cb3f329bca8cc13226e19fac34d25e4a0ecc Dereference SerializedPageReferences in batches (James Petty)
- 7ce6c1f1c80d3b39fce75a756bae47e63745ca96 Avoid unnecessary synchronization in LazyOutputBuffer (James Petty)
- ca9466572da066f29c894d6b2ee5600aeab2dffa Add listener model for TaskThresholdMemoryRevokingScheduler (Saksham Sachdev)
- ca7610b70370dbfa7b4a7cc7df6a03381e50cb36 Require min 30s query.min-expire-age (Ariel Weisberg)
- ab78e27d92221a616133027ef88b3385948ec8b3 Fix SliceDirectSelectiveStreamReader for ARRAY of VARCHAR (Ying Su)
- c3d0a4d762b387ab3c2f576e478a84527c9dc6d8 Fix testSubfieldValue in OrcTester (Ying Su)
- 27fc29b9a7d0b2e8704dfb3a42b00851535d1212 Limit batch read mode to certain conditions for SliceDirectSelectiveStreamReader (Ying Su)
- 70f379dd2b0287cdb4a92de59faadc71813d9a84 Fix SliceDirectSelectiveStreamReader for dataLength is 0 (Ying Su)
- 18cbfa78167df5d9e18172886d452a3bda4feda1 Extract assertBlockEquals() in OrcTester (Ying Su)
- 6f4e4443a21f82e1cd5014f641387dabc1e2623a Upgrade JTS to 1.18.0 (James Gill)
- cebe0c29aa14ce7c84d40c70684c3cc76bc990f9 Store manifest only when they are preferred and file renaming is enabled (Nikhil Collooru)
- aa1fd64545ee5be5bdfd0081a20107dad5b145a7 Track compressed manifest size when file_renaming is enabled (Nikhil Collooru)
- 6a47cf2557087b3d0aff4af154d8f4da53928297 Add containsNumberedFileNames to PartitionUpdate (Nikhil Collooru)
- f538920ae958a51961e0d1b2f134b40445844b5d Fix sorting file names for bucketed splits (Nikhil Collooru)
- effd0ea39af14f737b6b09d96df5f19d8fe66ddc Disallow ORDER BY literal (ordinal) in window functions as this may be a misunderstood feature. (Sreeni Viswanadha)
- 844dbebc2cd417835e4506fac31a7da600cf7e13 Avoid planning unnecessary Sort (George Wang)
- 90cd1d004cf458d397dd50c40bb57cc85ac53be9 Avoid planning unnecessary LIMIT/TopN/Sort/DistinctLimit (George Wang)
- b1846c1bae374a0bc70675166fa1903b2b1ddd83 Support driver level plan in fragment result caching (Shixuan Fan)
- b950aec7a5d4ebc12f56aad0a127f53c3601592c Remove unused method in LocalExecutionPlanner (Shixuan Fan)
- 8a005811dc02be9474c220a47c648a9c173b5e83 Provide more detail in MULTI_CATALOG_WRITE_CONFLICT (Wenlei Xie)
- abdbd68f56011b439f71e5657f56b721620b0ee0 Fix javadoc method reference (James Petty)
- 13be472fd60ef9a4d3a5a0b2620793515404e40a Fix typo in ConcurrentHashMap creation (James Petty)
- f5a5af4b6e9d741e4d893a407b53b3a2ee9bb757 Avoid repeatedly setting memory limits in QueryContext (James Petty)
- 1bb85b03c76ee564685d7dc0b6d32f5ce23b4aaf Fix QualifiedObjectName instantiation in FunctionAndTypeManager (Rongrong Zhong)
- f31b960020d39189c65220a8aa2fc41e5ff40568 Minor fix to Hive temporary table location (Wenlei Xie)
- adc7a640faf6263033fce8ea083ae3799f39ddce Avoid costly resizes of the addresses dynamic array in PagesIndex (Andrii Rosa)
- 9dd4b7766c461c3b717ea6636569733aee6e53be Remove redundant setter in TestHiveClientConfig (Shixuan Fan)
- 8d6ac0ee73f2bbc859f66e667c1b3a8ba99bac0a Reuse decompression buffer between multiple OrcInputStream's (Andrii Rosa)
- 47a39d4d5252da11aabab5bdc43c89d04e15d907 Shrink cached buffers in OrcInputStream (Andrii Rosa)
- 69b12d3bf1290cdc4bea01c6533dcab7999335c5 Improve OrcInputStream memory tracking (Andrii Rosa)
- 839007f9b78b8fc1159ea1c14f4f23ceddd57e54 Upgrade Drift to 1.32 (Andrii Rosa)
- bb3a2b78f97fd97b3ae005d5dbf8f46552309fcf Add PrestoThriftPage (Rongrong Zhong)
- 7561c121a28de44ad632cfbd28b189b04553fb0d Add flag to set cache eviction retries (Bin Fan)
- 7f3e6940da9cf187f30d8c79f5106d7b2c09db44 Improve error handling for spill (Rebecca Schlussel)
- 6d77b0a868647d93e035ca93d160aaab8d60cbc8 Make AbstractTestXQueries classes abstract (Rebecca Schlussel)
- 28f787bcd6b41ba2e2ff4bce4111018b3da2404d Add config to stop installation of disabled connectors (Saksham Sachdev)
- a68b96b06f24496ababb21018d82edd5091ee74b Split large pages in TempSingleStreamSpiller (Rebecca Schlussel)
- ab243b6ce24ac150bac4821c4a11342e89daa88a Refactor JoinBridgeProvider (Rebecca Schlussel)
- a2939dc3afea1d407a9777f027ab3da113d6dd4a Make getExpectedOrdersTableDescription protected (Rohit Jain)
- 382c83dc271617e7a33affb97228d2af5e84c2b4 Revert "Fix temporary table deletion for exchange materialization" (Wenlei Xie)
- cfe964254b14c684beeb6d795975a14842b1b704 Replace Joda-Time libraries with java.time in Presto MongoDB (Sorin Stoiana)
- 772621530a4b67f34f93bdeba9c29c86a4a5c978 Refactor AbstractTestQueryFramework to supply expectedQueryRunner (John Roll)
- 1acb585f0195315fc55dc9c6e83d4405ef06fb30 Release notes for 0.244.1 (Bhavani Hari)
- d2cb24cd4cec630940463d8d8dc25806c774f00d Fix temporary table deletion for exchange materialization (shenh062326)
- 25c03ccad1758ee2a3d0336e73e49874e7f29ede Fix accidental per-instance logger in SerializedPageWriteListener (James Petty)
- a4d7abded6aefddbe3526ad8ee0f01bbc55cb51f Set broadcast memory limit for Presto on Spark (Vic Zhang)
- 3feecf3743ffd8c81c21b08ed89402ca45fe8c05 Fix broadcast memory update in RootAggregatedMemoryContext (Vic Zhang)
- 4d5afceed18cc4f71a1574adafcd506db6e33e83 Support Serialized Presto Page in Thrift UDF service (Rongrong Zhong)
- b49294b10a72654d6bc5a1f44d88979e1fd2f286 Move BlockEncodingManager to presto-common (Rongrong Zhong)
- b3d92fcef735b6cce29ade81a452d5b5ba968dd6 Remove redundant code in EchoFirstInputThriftUdfService (Rongrong Zhong)
- 8bea180fdc31be1f2c31b9e9d2b3b6f83ec47c63 Prefer default parallelism to avoid adding extra local exchange (Andrii Rosa)
- 1eb7024d34b597367a6c4be071a8cbe0e35418bf Support different concurrency settings for table writer (Andrii Rosa)
- b7d1a76eb2d443f65ada17357e169e7bdda6c8be Add reverse function for varbinary (Emy Sun)
- 22efd1acfdf0f1c33aa8e3a3a2a9ead87905b8e8 Deprecating Joda library in Presto Cassandra (Moji Solgi)
- 222b0df79bd9c0dc2799451ddb49c15643d8bfed Inject null provider when thrift execution is not needed (Rongrong Zhong)
- 60f4d6891e28cb78665a4e9f736977b19623bae7 Fix incompatible reflection for modifying static final fields since JDK 12 (fornaix)
- bf7bec4438accdde7bd60ec94845de7a1696be4b handle nulls properly in flatmap reader (Zhenyuan Zhao)
- 9b3690acaf956a4e8f79dc2d6b504078400c9079 Clean up parameters related code (Rebecca Schlussel)
- 457aa86c620f150a0837ddc0cdb456473c991c32 Only allow constants in prepared statement parameters (Rebecca Schlussel)
- 729554d3a5b83909322417703992f8cc5e764497 Fix dynamic pruning for null keys in hive partition (George Wang)
- 6913ec9a25187748dea95f991a139667d876526a Minor fix for memory property removal in Presto verifier (Wenlei Xie)

### CODE REVIEW

**Comment by rschlussel** [comment]
Additional release notes for me:
Optimize queries that have unnecessary limits and order bys.

**Comment by rschlussel** [comment]
@rongrong  i think we're still missing a release note for  "#15508 Port reverse(varbinary) from Prestosql to Prestodb (Merged by: Rongrong Zhong)"

**Comment by rschlussel** [comment]
@caithagoras  can't add you as a reviewer because the PR is under your name, but the release notes are ready for review.

File: presto-docs/src/main/sphinx/release/release-0.246.rst [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+**Details**
+===========
+
+General Changes
+_______________
+* Fix a bug introduced in :pr:`15313` that would cause queries to fail when using upper case in SQL function catalog schema names.
+* Fix a possible int overflow error when spilling to temp storage.
+* Fix an issue where prepared statements would allow some non-constant parameters.
+* Fix injection issue when using function namespace manager.
+* Add 30 second minimum to `query.min-expire-age` configuration parameter.
```
Comment (caithagoras) [inline presto-docs/src/main/sphinx/release/release-0.246.rst:17]:
```
* Add a minimum value of 30 seconds to the configuration property ``query.min-expire-age``.
```

File: presto-docs/src/main/sphinx/release/release-0.246.rst [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+* Fix a bug introduced in :pr:`15313` that would cause queries to fail when using upper case in SQL function catalog schema names.
+* Fix a possible int overflow error when spilling to temp storage.
+* Fix an issue where prepared statements would allow some non-constant parameters.
+* Fix an issue where prepared statements would allow some non-constant parameters.
+* Fix injection issue when using function namespace manager.
+* Add 30 second minimum to `query.min-expire-age` configuration parameter.
+* Add listener-based revocation model for spilling strategy PER_TASK_MEMORY_THRESHOLD.
+* Disable spill to disk for join queries where the probe side uses grouped execution and the build does not.  Previously these queries would fail with generic_internal_errors.
+* Disallow ORDER BY literals used with Window functions as it's not useful, expensive and most often used wrongly.
+* Support fragment result caching for queries with local exchange (e.g. intermediate aggregation).
```
[the remark this answers is not in the document]
Comment (caithagoras) [reply presto-docs/src/main/sphinx/release/release-0.246.rst:21]: Guideline for the start of the line: https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines#order-of-changes.

We should rephrase to
```
Add support ...
```

File: presto-docs/src/main/sphinx/release/release-0.246.rst [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+General Changes
+_______________
+* Fix a bug introduced in :pr:`15313` that would cause queries to fail when using upper case in SQL function catalog schema names.
+* Fix a possible int overflow error when spilling to temp storage.
+* Fix an issue where prepared statements would allow some non-constant parameters.
+* Fix injection issue when using function namespace manager.
+* Add 30 second minimum to `query.min-expire-age` configuration parameter.
+* Add listener-based revocation model for spilling strategy PER_TASK_MEMORY_THRESHOLD.
+* Disable spill to disk for join queries where the probe side uses grouped execution and the build does not.  Previously these queries would fail with generic_internal_errors.
+* Disallow ORDER BY literals used with Window functions as it's not useful, expensive and most often used wrongly.
```
Comment (caithagoras) [inline presto-docs/src/main/sphinx/release/release-0.246.rst:20]: double back quotes. Please also update everywhere else in the doc for keywords.
```
``ORDER BY``
```

File: presto-docs/src/main/sphinx/release/release-0.246.rst [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+General Changes
+_______________
+* Fix a bug introduced in :pr:`15313` that would cause queries to fail when using upper case in SQL function catalog schema names.
+* Fix a possible int overflow error when spilling to temp storage.
+* Fix an issue where prepared statements would allow some non-constant parameters.
+* Fix injection issue when using function namespace manager.
+* Add 30 second minimum to `query.min-expire-age` configuration parameter.
+* Add listener-based revocation model for spilling strategy PER_TASK_MEMORY_THRESHOLD.
+* Disable spill to disk for join queries where the probe side uses grouped execution and the build does not.  Previously these queries would fail with generic_internal_errors.
```
Comment (caithagoras) [inline presto-docs/src/main/sphinx/release/release-0.246.rst:19]: Remove extra whitespace.
```
``GENERIC_INTERNAL_ERRORS``
```

File: presto-docs/src/main/sphinx/release/release-0.246.rst [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+* Disallow ORDER BY literals used with Window functions as it's not useful, expensive and most often used wrongly.
+* Support fragment result caching for queries with local exchange (e.g. intermediate aggregation).
+* Optimize queries that have unnecessary limits and order bys.
+  This feature is enabled by default and can be disabled by using the configuration property optimizer.skip-redundant-sort or session property skip_redundant_sort
```
Comment (caithagoras) [inline presto-docs/src/main/sphinx/release/release-0.246.rst:23]: Add a period (".") at the end of the line.

File: presto-docs/src/main/sphinx/release/release-0.246.rst [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+* Fix a bug introduced in :pr:`15313` that would cause queries to fail when using upper case in SQL function catalog schema names.
+* Fix a possible int overflow error when spilling to temp storage.
+* Fix an issue where prepared statements would allow some non-constant parameters.
+* Fix injection issue when using function namespace manager.
+* Add 30 second minimum to `query.min-expire-age` configuration parameter.
+* Add listener-based revocation model for spilling strategy PER_TASK_MEMORY_THRESHOLD.
+* Disable spill to disk for join queries where the probe side uses grouped execution and the build does not.  Previously these queries would fail with generic_internal_errors.
+* Disallow ORDER BY literals used with Window functions as it's not useful, expensive and most often used wrongly.
+* Support fragment result caching for queries with local exchange (e.g. intermediate aggregation).
+* Optimize queries that have unnecessary limits and order bys.
```
Comment (caithagoras) [inline presto-docs/src/main/sphinx/release/release-0.246.rst:22]: We should rephrase to
```
Improve ...
```

File: presto-docs/src/main/sphinx/release/release-0.246.rst [hunk 7]
```diff
@@ -0,0 +1,36 @@
+=============
+Release 0.246
+=============
+
+**Highlights**
+==============
+
```
Comment (caithagoras) [inline presto-docs/src/main/sphinx/release/release-0.246.rst:7]: Let's include all changes that starts with "Add" and "Improve".

File: presto-docs/src/main/sphinx/release/release-0.246.rst [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+**Details**
+===========
+
+General Changes
+_______________
+* Fix a bug introduced in :pr:`15313` that would cause queries to fail when using upper case in SQL function catalog schema names.
+* Fix a possible int overflow error when spilling to temp storage.
```
Comment (mayankgarg1990) [inline presto-docs/src/main/sphinx/release/release-0.246.rst:14]: `int` -> `integer`, `temp` -> `temporary`

File: presto-docs/src/main/sphinx/release/release-0.246.rst [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+* Fix an issue where prepared statements would allow some non-constant parameters.
+* Fix injection issue when using function namespace manager.
+* Add 30 second minimum to `query.min-expire-age` configuration parameter.
+* Add listener-based revocation model for spilling strategy PER_TASK_MEMORY_THRESHOLD.
```
Comment (mayankgarg1990) [inline presto-docs/src/main/sphinx/release/release-0.246.rst:18]: what does this mean - can the author clarify this ?

File: presto-docs/src/main/sphinx/release/release-0.246.rst (hunk 5 above)
Comment (mayankgarg1990) [reply presto-docs/src/main/sphinx/release/release-0.246.rst:23]: back quote properties

File: presto-docs/src/main/sphinx/release/release-0.246.rst (hunk 7 above)
Comment (mayankgarg1990) [reply presto-docs/src/main/sphinx/release/release-0.246.rst:7]: I don't think there is anything that major. lets remove this section for this release

**Review by mayankgarg1990 — CHANGES_REQUESTED** [review]

**Comment by rschlussel** [comment]
@cocozianu can you squash all the commits together?

File: presto-docs/src/main/sphinx/release/release-0.246.rst [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+**Details**
+===========
+
+General Changes
+_______________
+* Fix a bug introduced in :pr:`15313` that would cause queries to fail when using upper case in SQL function catalog schema names.
+* Fix a possible int overflow error when spilling to temp storage.
+* Fix an issue where prepared statements would allow some non-constant parameters.
+* Fix an issue where prepared statements would allow some non-constant parameters.
+* Fix injection issue when using function namespace manager.
```
[the remark this answers is not in the document]
Comment (caithagoras) [reply presto-docs/src/main/sphinx/release/release-0.246.rst:13]: This comment has not been addressed yet.

**Comment by caithagoras** [comment]
Please squash all commits into 1. We don't use multiple commits for retaining history.

File: presto-docs/src/main/sphinx/release/release-0.246.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (rongrong) [reply presto-docs/src/main/sphinx/release/release-0.246.rst:13]: It means if user use function namespace manager presto server might fail to start failing with injection error.

**Review by rongrong — APPROVED** [review]

**Pull request merged**